.github/workflows/build_charms.yaml

```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+name: Build Charms
+
+on:
+  workflow_call:
+    inputs:
+      runs-on:
+        type: string
+        description: Image runner for building the images
+        default: ubuntu-22.04
+      working-directory:
+        type: string
+        description: The working directory for jobs
+        default: "./"
+    outputs:
+      charms:
+        description: List of charms built
+        value: ${{ jobs.get-charms.outputs.charms }}
+
+jobs:
+  get-charm-names:
+    name: Get charms
+    runs-on: ${{ inputs.runs-on }}
+    outputs:
+      charms: ${{ env.CHARM_NAMES }}
+    steps:
+      - uses: actions/checkout@v4.1.1
+      - name: Find charm paths
+        working-directory: ${{ inputs.working-directory }}
+        run: |
+          echo "CHARM_PATHS=$(find * -type f -name metadata.yaml 2> /dev/null |
+          sed 's:[^/]*$::' |
+          jq -Rsc '. / "\n" - [""]')" >> $GITHUB_ENV
+      - name: Extract charm names
+        working-directory: ${{ inputs.working-directory }}
+        run: |
+          echo "CHARM_NAMES=$(for path in $(jq -cr '.[]' ); do
+            echo {\"name\":\"$(yq '.name' $path/metadata.yaml)\", \"path\":\"$path\"}
+          done <<< $CHARM_PATHS |
+          jq -sc '.')" >> $GITHUB_ENV
+  build-charms:
+    name: Build and push charms
+    runs-on: ${{ inputs.runs-on }}
+    needs: [get-charm-names]
+    strategy:
+      matrix:
+        include: ${{ fromJSON(needs.get-charm-names.outputs.charms) }}
+    steps:
+      - uses: actions/checkout@v4.1.1
+      - uses: canonical/setup-lxd@v0.1.1
+      - name: Pack charm
+        if: ${{ !cancelled() }}
+        working-directory: ${{ inputs.working-directory }}/${{ matrix.path }}
+        run: |
+          sudo snap install charmcraft --classic --channel latest/stable
+          charmcraft pack -v
+          echo "CHARM_FILE=$(ls ${{matrix.name}}_*.charm)" >> $GITHUB_ENV
+      - name: Upload charm artifact
+        if: ${{ !cancelled() }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: charm-artifacts
+          path: ${{ inputs.working-directory }}/${{ matrix.path }}${{ env.CHARM_FILE }}
+          if-no-files-found: error
```

Comment:
I think artifacts with the same name will overwrite each other? So  I think the matrix here will only save one charm as artifact.

Reply:
correct, artifacts with the same name would clash.

I am attempting to push charms like this into the file store
`aaa_ubuntu-22.04-amd64.charm`
`bbb_ubuntu-22.04-amd64.charm`

if two or more sub-directories had charms whose resultant filenames were the same (ie -- two charm file both with the name `ccc__ubuntu-22.04-amd64.charm`) then only one would be stored in the artifact for the action run.  I'm not entirely sure how to solve that -- even `${{ env.CHARM_NAME }}-charm` doesn't address this so it's no worse than the existing implementation

Reply:
Didn't know GitHub workflow artifacts is merged when upload multiple times, thanks for the information!

Reply:
https://github.com/actions/upload-artifact#usage

oh sure, workflow artifact uploading is cool.
* individual files
* entire directories
* wildcard pattern uploads
* multiple-paths with file matching exclusions